cpp/src/interop/to_arrow_host.cu

```diff
@@ -0,0 +1,429 @@
+  rmm::cuda_stream_view stream;
+  rmm::device_async_resource_ref mr;
+
+  int populate_validity_bitmap(ArrowBitmap* bitmap) const
+  {
+    if (!column.has_nulls()) { return NANOARROW_OK; }
+
+    NANOARROW_RETURN_NOT_OK(ArrowBitmapResize(bitmap, static_cast<int64_t>(column.size()), 0));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(bitmap->buffer.data,
+                                  (column.offset() > 0)
+                                    ? cudf::detail::copy_bitmask(column, stream, mr).data()
+                                    : column.null_mask(),
+                                  bitmap->buffer.size_bytes,
+                                  cudaMemcpyDefault,
+                                  stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T>
+  int populate_data_buffer(device_span<T const> input, ArrowBuffer* buffer) const
+  {
+    NANOARROW_RETURN_NOT_OK(ArrowBufferResize(buffer, input.size_bytes(), 1));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+      buffer->data, input.data(), input.size_bytes(), cudaMemcpyDefault, stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(!is_rep_layout_compatible<T>() &&
+                           (!std::is_same_v<T, numeric::decimal128> &&
+                            !std::is_same_v<T, numeric::decimal32> &&
+                            !std::is_same_v<T, numeric::decimal64>))>
+  int operator()(ArrowArray*) const
+  {
+    CUDF_FAIL("Unsupported type for to_arrow_host", cudf::data_type_error);
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(is_rep_layout_compatible<T>() || std::is_same_v<T, numeric::decimal128>)>
+  int operator()(ArrowArray* out) const
+  {
+    nanoarrow::UniqueArray tmp;
+
+    auto const storage_type = id_to_arrow_storage_type(column.type().id());
+    NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), storage_type, column));
+
+    NANOARROW_RETURN_NOT_OK(populate_validity_bitmap(ArrowArrayValidityBitmap(tmp.get())));
+    using DataType = std::conditional_t<std::is_same_v<T, numeric::decimal128>, __int128_t, T>;
+    NANOARROW_RETURN_NOT_OK(
+      populate_data_buffer(device_span<DataType const>(column.data<DataType>(), column.size()),
+                           ArrowArrayBuffer(tmp.get(), fixed_width_data_buffer_idx)));
+
+    ArrowArrayMove(tmp.get(), out);
+    return NANOARROW_OK;
+  }
+
+  // convert decimal types from libcudf to arrow where those types are not directly
+  // supported by Arrow. These types must be fit into 128 bits, the smallest
+  // decimal resolution supported by Arrow
+  template <typename T,
+            CUDF_ENABLE_IF(not is_rep_layout_compatible<T>() &&
+                           (std::is_same_v<T, numeric::decimal32> ||
+                            std::is_same_v<T, numeric::decimal64>))>
+  int operator()(ArrowArray* out) const
+  {
+    using DeviceType = std::conditional_t<std::is_same_v<T, numeric::decimal32>, int32_t, int64_t>;
+    nanoarrow::UniqueArray tmp;
```

Comment:
The variable name `tmp` could be improved to a more descriptive one such as ``arrow_array` or `arrow_data_array` across the file

Reply:
Let's leave it as is. This is the pattern also used throughout all of libcudf's usage of nanoarrow now so we're matching the convention, as it were.